File: src/infra/auth/session.ts
/**
 * Session → `TenantContext` resolution (§5.1).
 *
 * Every use case takes a `TenantContext` built here, **server-side, from the
 * session**. A `tenantId` arriving in a request body is an attacker's
 * suggestion, not a fact, and nothing in this file ever reads one.
 *
 * `getClaims()` rather than `getSession()`: it verifies the JWT signature and
 * returns the custom `tenant_id` / `user_role` claims injected by
 * `custom_access_token_hook`. `getSession()` returns whatever is in the cookie
 * without verifying it, which is precisely the value an attacker controls.
 */
import "server-only";
import { cache } from "react";
import { headers } from "next/headers";
import type { SupabaseClient } from "@supabase/supabase-js";
import type { TenantContext } from "@/core/domain/tenant-context";
import type { UserRole } from "@/core/domain/enums";
import { createClient } from "../supabase/server";
import { createBearerClient } from "../supabase/bearer";
import { firstRelated } from "../supabase/mappers";
import { parseBearerToken } from "@/lib/bearer-token";
import type { Database } from "../supabase/database.types";

export interface SessionUser extends TenantContext {
  /** Gates every route until the user chooses their own password (D-02). */
  readonly mustChangePassword: boolean;
  readonly fullName: string;
  readonly profileStatus: "ACTIVE" | "DISABLED";
  /** The mess's own name, shown to its members in place of ours. */
  readonly tenantName: string;
  /** Storage path of the mess's logo, if it has uploaded one. */
  readonly tenantLogoPath: string | null;
}

/**
 * The current user, or null when unauthenticated.
 *
 * Wrapped in React `cache` so that a layout, a page and several components in
 * one render share a single database round trip rather than issuing one each.
 * The cache is per-request, so it cannot leak one user's context into another's.
 */
export const getSessionUser = cache(async (): Promise<SessionUser | null> => {
  // A bearer token wins when one is present, so **every** existing caller —
  // pages, Server Actions and route handlers alike — serves the mobile app
  // without being rewritten. There are 115 of them; making the chokepoint
  // transport-aware is the only version of this that does not mean touching
  // each one and getting one wrong.
  //
  // A browser never sends this header, so the cookie path is untouched. And a
  // header alone grants nothing: the token below is signature-verified, so
  // inventing one gets you no further than inventing a cookie would.
  const token = parseBearerToken((await headers()).get("authorization"));
  if (token) return getSessionUserFromToken(token);

  const supabase = await createClient();

  const { data, error } = await supabase.auth.getClaims();
  if (error || !data?.claims) return null;

  const claims = data.claims as {
    sub?: string;
    tenant_id?: string;
    user_role?: string;
  };
  if (!claims.sub) return null;

  return resolveSessionUser(supabase, claims.sub);
});

/**
 * The same user, resolved from an `Authorization: Bearer` token instead of a
 * cookie — the mobile app's transport.
 *
 * Returns the **identical `SessionUser`**, so every use case, policy and route
 * handler behind this function is transport-agnostic and neither knows nor
 * cares which client is calling.
 *
 * The token is verified, not trusted: `getClaims(accessToken)` checks the
 * signature against the project's JWKS exactly as the cookie path does. An
 * expired or forged token yields no claims and this returns null.
 *
 * Cached per token rather than per request. `cache()` keys on arguments, so two
 * different callers in one render cannot collide — but note the key is the
 * token itself, which is what makes that safe.
 */
export const getSessionUserFromToken = cache(
  async (accessToken: string): Promise<SessionUser | null> => {
    const supabase = createBearerClient(accessToken);

    const { data, error } = await supabase.auth.getClaims(accessToken);
    if (error || !data?.claims) return null;

    const claims = data.claims as { sub?: string };
    if (!claims.sub) return null;

    return resolveSessionUser(supabase, claims.sub);
  },
);

/**
 * Claims → `SessionUser`, shared by both transports.
 *
 * Kept in one place deliberately: the fail-closed rules below are the whole
 * authorization story for a suspended tenant or a disabled account, and a
 * second copy that drifted would be a silent way back in.
 */
async function resolveSessionUser(
  supabase: SupabaseClient<Database>,
  sub: string,
): Promise<SessionUser | null> {
  // Profile, tenant and student in one round trip. This runs on every
  // authenticated request, so a second query here would be felt everywhere.
  const { data: profile, error: profileError } = await supabase
    .from("profiles")
    .select(
      `id, tenant_id, role, full_name, status, must_change_password,
       tenants!inner ( slug, name, timezone, status, logo_path ),
       students ( id )`,
    )
    .eq("id", sub)
    .maybeSingle();

  // Fail closed (§2.7). A session whose profile cannot be read is
  // indeterminate, and indeterminate means unauthenticated.
  if (profileError || !profile) return null;

  const tenant = firstRelated<{
    slug: string;
    name: string;
    timezone: string;
    status: string;
    logo_path: string | null;
  }>(profile.tenants as never);
  if (!tenant) return null;

  // A suspended tenant or a disabled account must not hold a usable session,
  // even though their JWT is still cryptographically valid until it expires.
  if (tenant.status !== "ACTIVE") return null;
  if (profile.status !== "ACTIVE") return null;

  // PostgREST collapses this embed to a single OBJECT, not an array, because
  // `students.profile_id` is unique. Reading `students[0]` silently yielded
  // undefined and left every student session without a studentId — which the
  // QR endpoint then refused as FORBIDDEN. See firstRelated().
  const studentId = firstRelated<{ id: string }>(profile.students as never)?.id;

  return {
    tenantId: profile.tenant_id,
    tenantSlug: tenant.slug,
    tenantName: tenant.name,
    tenantLogoPath: tenant.logo_path,
    timezone: tenant.timezone,
    actorProfileId: profile.id,
    role: profile.role as UserRole,
    ...(studentId ? { studentId } : {}),
    mustChangePassword: profile.must_change_password,
    fullName: profile.full_name,
    profileStatus: profile.status,
  };
}

/**
 * The current user, or throws.
 *
 * For code paths already behind route gating, where "no session" means the
 * gating is broken rather than a case to render.
 */
export async function requireSessionUser(): Promise<SessionUser> {
  const user = await getSessionUser();
  if (!user) throw new Error("No authenticated session");
  return user;
}

/** The landing route for a role. One place, so gating and redirects agree. */
export function homeRouteFor(role: UserRole): string {
  switch (role) {
    case "STUDENT":
      return "/student";
    case "STAFF":
      return "/staff";
    case "ADMIN":
    case "SUPER_ADMIN":
      return "/admin";
  }
}

File: mobile/lib/src/data/session.dart
/// The signed-in user, exactly as `GET /api/me` and `POST /api/auth/login`
/// describe them.
///
/// Mirrors `toSessionPayload` on the server, which deliberately omits
/// `tenantId`, `actorProfileId` and `studentId`. The app never needs them: every
/// endpoint rebuilds its own tenant context from the token, and an id the client
/// could send back is the multi-tenancy bug rule 8 exists to prevent. If a field
/// here ever needs one of those, the answer is a server-side change, not a
/// wider payload.
library;

enum UserRole {
  student,
  staff,
  admin,
  superAdmin;

  static UserRole fromWire(String value) => switch (value) {
    'STUDENT' => UserRole.student,
    'STAFF' => UserRole.staff,
    'ADMIN' => UserRole.admin,
    'SUPER_ADMIN' => UserRole.superAdmin,
    // Fail closed (rule 7): an unrecognised role is treated as the least
    // privileged one rather than waved through. A newer server that adds a role
    // must not be able to grant this build a shell it does not understand.
    _ => UserRole.student,
  };

  /// Admins operate a counter during a rush, and the audit trail records who
  /// did. This mirrors `ROLE_GATES` in the web proxy exactly.
  bool get usesStaffShell =>
      this == UserRole.staff ||
      this == UserRole.admin ||
      this == UserRole.superAdmin;
}

class Session {
  const Session({
    required this.role,
    required this.fullName,
    required this.mustChangePassword,
    required this.tenantSlug,
    required this.tenantName,
    required this.tenantLogoUrl,
    required this.timezone,
    required this.isStudent,
  });

  final UserRole role;
  final String fullName;

  /// Gates every screen until the user chooses their own password (D-02).
  final bool mustChangePassword;

  final String tenantSlug;

  /// The mess's own name. Shown throughout the app in place of ours — once
  /// somebody has signed in they are inside *their hostel's* app.
  final String tenantName;

  /// Route to the mess's logo, or null if it has not uploaded one, in which
  /// case its name stands in.
  final String? tenantLogoUrl;

  /// IANA zone of the **mess**, not the device. Every date rendered anywhere in
  /// the app derives from this — a student travelling, or a phone with its
  /// clock set wrong, must still see the mess's own service dates (rule 9).
  final String timezone;

  final bool isStudent;

  factory Session.fromJson(Map<String, dynamic> json) => Session(
    role: UserRole.fromWire(json['role'] as String? ?? 'STUDENT'),
    fullName: json['fullName'] as String? ?? '',
    mustChangePassword: json['mustChangePassword'] as bool? ?? false,
    tenantSlug: json['tenantSlug'] as String? ?? '',
    tenantName: json['tenantName'] as String? ?? '',
    tenantLogoUrl: json['tenantLogoUrl'] as String?,
    timezone: json['timezone'] as String? ?? 'Asia/Kolkata',
    isStudent: json['isStudent'] as bool? ?? false,
  );

  Session copyWith({bool? mustChangePassword}) => Session(
    role: role,
    fullName: fullName,
    mustChangePassword: mustChangePassword ?? this.mustChangePassword,
    tenantSlug: tenantSlug,
    tenantName: tenantName,
    tenantLogoUrl: tenantLogoUrl,
    timezone: timezone,
    isStudent: isStudent,
  );
}

/// The tokens themselves, kept apart from [Session] because they go to secure
/// storage while the session is only ever held in memory.
class AuthTokens {
  const AuthTokens({required this.accessToken, required this.refreshToken});

  final String accessToken;
  final String refreshToken;
}
